# **IRONHACK_Sharks**

In [27]:
import pandas as pd
import numpy as np
import warnings
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
warnings.filterwarnings('ignore')  
pd.set_option('display.max_columns', None)  
from src.shark_func import *

Proyecto de limpieza del archivo attacks.csv, que se puede localizar en el siguiente link https://www.kaggle.com/datasets/teajay/global-shark-attacks/download?datasetVersionNumber=7

In [2]:
# attacks.csv no está codificado en UTF-8, hay que especificar la codificación en la importación
sharks = pd.read_csv('../data/attacks.csv',encoding='ANSI')    

In [3]:
shape_origin = sharks.shape         # Guardamos el tamaño inicial del dataset
sharks.head(15)

Case Number         Date    Year        Type    Country   
0     2018.06.25  25-Jun-2018  2018.0     Boating        USA  \
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
5   2018.06.03.b  03-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
6   2018.06.03.a  03-Jun-2018  2018.0  Unprovoked     BRAZIL   
7     2018.05.27  27-May-2018  2018.0  Unprovoked        USA   
8   2018.05.26.b  26-May-2018  2018.0  Unprovoked        USA   
9   2018.05.26.a  26-May-2018  2018.0  Unprovoked        USA   
10    2018.05.24  24-May-2018  2018.0    Provoked  AUSTRALIA   
11    2018.05.21  21-May-2018  2018.0  Unprovoked        USA   
12  2018.05.13.b  13-May-2018  2018.0  Unprovoked        USA   
13  2018.05.13.a  13-May-2018  2018.0     Invalid    ENGLAND   
14    2018.05.00     May 2018  2018.0    Provoked  AUSTRALIA   

                 Area                                           Location   
0          California                        Oceanside, San Diego County  \
1             Georgia                     St. Simon Island, Glynn County   
2              Hawaii                                       Habush, Oahu   
3     New South Wales                                 Arrawarra Headland   
4              Colima                                           La Ticla   
5     New South Wales                                 Flat Rock, Ballina   
6          Pernambuco                              Piedade Beach, Recife   
7             Florida  Lighhouse Point Park, Ponce Inlet, Volusia County   
8             Florida                       Cocoa Beach, Brevard  County   
9             Florida                      Daytona Beach, Volusia County   
10         Queensland                                    Cairns Aquarium   
11     South Carolina                  Isle of Palms,  Charleston County   
12     South Carolina                Hilton Head Island, Beaufort County   
13           Cornwall                                     Off Land's End   
14  Westerm Australia                                         Dugong Bay   

           Activity                    Name Sex   Age   
0          Paddling             Julie Wolfe    F   57  \
1          Standing         Adyson McNeely     F   11   
2           Surfing             John Denges    M   48   
3           Surfing                    male    M  NaN   
4       Free diving          Gustavo Ramos     M  NaN   
5      Kite surfing                 Chris …    M  NaN   
6          Swimming  Jose Ernesto da Silva     M   18   
7           Fishing                    male    M   52   
8           Walking               Cody High    M   15   
9          Standing                    male    M   12   
10   Feeding sharks                    male    M   32   
11  Boogie boarding            Trey de Boer    M   10   
12         Swimming             Jei Turrell    M   10   
13          Fishing            Max Berryman    M   21   
14   Feeding sharks         Melisa Brunning    F   34   

                                               Injury Fatal (Y/N)   
0   No injury to occupant, outrigger canoe and pad...           N  \
1                          Minor injury to left thigh           N   
2        Injury to left lower leg from surfboard skeg           N   
3                           Minor injury to lower leg           N   
4   Lacerations to leg & hand shark PROVOKED INCIDENT           N   
5                             No injury, board bitten           N   
6                                               FATAL           Y   
7             Minor injury to foot. PROVOKED INCIDENT           N   
8                               Lower left leg bitten           N   
9                                Minor injury to foot           N   
10  Minor bite to hand by captive shark. PROVOKED ...           N   
11               Injur

In [4]:
sharks.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

Vamos a comprobar si tenemos filas duplicadas

In [5]:
sharks.duplicated().sum()

19411

Tenemos 19411 lineas duplicadas. Procedemos a eliminarlas. No aportan información alguna

In [6]:
sharks.drop_duplicates(inplace = True)
sharks.reset_index(drop=True,inplace=True)

In [7]:
sharks.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6312 entries, 0 to 6311
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             6310 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null   

In [8]:
sharks_info = cols_info(sharks)  # La función de cols_info(). Se puede encontrar en el archivo shark_func, ubicado en la carpeta src
                                 # Da información de los tipos de dato de cada columna
    
sharks_info

Col Type Nulos   str float int bool date float==nan   
Case Number              object     2  6310     2   0    0    0       True  \
Date                     object    10  6302    10   0    0    0       True   
Year                    float64    12     0  6312   0    0    0      False   
Type                     object    14  6298    14   0    0    0       True   
Country                  object    60  6252    60   0    0    0       True   
Area                     object   465  5847   465   0    0    0       True   
Location                 object   550  5762   550   0    0    0       True   
Activity                 object   554  5758   554   0    0    0       True   
Name                     object   220  6092   220   0    0    0       True   
Sex                      object   575  5737   575   0    0    0       True   
Age                      object  2841  3471  2841   0    0    0       True   
Injury                   object    38  6274    38   0    0    0       True   
Fatal (Y/N)              object   549  5763   549   0    0    0       True   
Time                     object  3364  2948  3364   0    0    0       True   
Species                  object  2848  3464  2848   0    0    0       True   
Investigator or Source   object    27  6285    27   0    0    0       True   
pdf                      object    10  6302    10   0    0    0       True   
href formula             object    11  6301    11   0    0    0       True   
href                     object    10  6302    10   0    0    0       True   
Case Number.1            object    10  6302    10   0    0    0       True   
Case Number.2            object    10  6302    10   0    0    0       True   
original order          float64     3     0  6312   0    0    0      False   
Unnamed: 22              object  6311     1  6311   0    0    0       True   
Unnamed: 23              object  6310     2  6310   0    0    0       True   

                       unique  
Case Number              6288  
Date                     5434  
Year                      250  
Type                        9  
Country                   213  
Area                      826  
Location                 4109  
Activity                 1533  
Name                     5231  
Sex                         7  
Age                       158  
Injury                   3738  
Fatal (Y/N)                 9  
Time                      367  
Species                  1550  
Investigator or Source   4970  
pdf                      6292  
href formula             6291  
href                     6286  
Case Number.1            6286  
Case Number.2            6287  
original order           6309  
Unnamed: 22                 2  
Unnamed: 23                 3

### PRIMERAS IMPRESIONES:

Solo quitando líneas duplicadas hemos reducido significativamente el tamaño del dataset. Pasando de 25000 registros a poco más de 6000

Vemos que todas las columnas tienen valores nulos, por lo que el tipo de dato de cada columna no nos aporta gran información de momento porque la presencia de nulos, hace que el tipo de la columna sea float o object (string)

* **Case Number:** identificación única del caso. Parece que están indentificadas con la fecha, y que están en orden decreciente. Vemos que los 
* **Date:** Fecha del registro, parece la misma que la columna 'Case Number'.
* **Year:** Año del suceso. Estas 3 columnas pueden darnos información valiosa a la hora de sustituir nulos, ya que la información de una la podamos aplicar a las otras. Veo que tenemos 250 valores únicos, de primeras esto parece extraño, ya que es bastante imporbable que tengamos registros de 250 años
* **Type:**  Tiene pinta que es tipo de ataque. Tendrá que ser una variable categórica ya que si asignamos un número a un ataque estaremos introduciendo un sesgo en los datos. Solo tiene 9 cateogorías únicas
* **Country:** País del ataque. Tiene 60 valores nulos, el resto string, lo qué es coherente. 213 valores únicos.
* **Area:** Area detro del pais (Estado/provincia/ciudad). Tiene 465 valores nulos, el resto string, lo qué es coherente. 826 valores únicos. De primeras parece una columna bastante complicada de formatear
* **Location:** Localización más precisa. Tiene 550 valores nulos, el resto string, lo qué es coherente. 4109 valores únicos. Columna compleja de formatear. Esta columna puede darnos información para completar información en Area y viceversa
* Name 
* Sex
* Age
* Injury
* Fatal (yes or no/ Unknown)
* Time
* Species (aquí hay chica)
* Investigator or Source: posibilidad de pasarla a dos columnas
* pdf
* href formula
* href duplicada
* Case number 1
* Case number 2
* original order: WTF
* Unnamed 22: Las voy a usar para otras cosas
* Unnamed 23: Las voy a usar para otras cosas

In [ ]:
len(check_nan_cols(sharks))

In [ ]:
len(check_nan_cols(sharks,method='avg'))

### **1 - Eliminación de nulos**

Nuestro primer objetivo es que no haya ninguna columna con valores nulos. Se observa que todas las columnas tienen nulos. En algunas (Unnamed) prácticamente el 100%. Iremos analizando cada una, e iremos tomando decisiones para ir resolviendo los conflictos

#### **Columnas Unnamed**

In [ ]:
# Hemos visto que solo hay 2 valores no nulos en Unnamed:23 y un valor no nulo en Unnamed:22

sharks[(sharks['Unnamed: 22'].notna()) | sharks['Unnamed: 23'].notna()]   # devuleve las filas en las que tengamos nulos en las columnas Unnamed 

Los valores no nulos que tenemos son: 'stopped here', 'Terrano' y 'change filename'. Información sin realción alguna con los ataques a tiburones. Vamos a optar añadir a estas dos columnas por 'Unknown'. Más adelante veremos si nos es conveniente usar estas columnas para albergar otra información que consideremos relevante


In [ ]:
sharks[['Unnamed: 22', 'Unnamed: 23']] = ['Unknown','Unknown']

In [ ]:
# Vuelvo a checkear duplicados en las filas

sharks.duplicated().sum()

In [ ]:
check_nan_cols(sharks,);

#### **Columna original order**

In [ ]:
sharks['original order'].head(10)

In [ ]:
sharks['original order'].tail(10)

In [ ]:
sharks['original order'].value_counts()  # Columna de enteros en la que todos son valores únicos menos 569 que se repite 1 vez

In [ ]:
sharks[sharks['original order']==569]

In [ ]:
sharks['original order'].max() - sharks['original order'].min() 

Esta columna representa el numero de ordenación original del dataset. Hemos llegado a la siguiente concluisión con estos datos:
* Vemos que registros consecutivos tienen numeros consecutivos en esta columna
* La diferencia entre el maximo y el minomo es de 6308, si le sumamos el valor repetido y los tres valores nulos suman 6312. La longitud de nuestro dataset
* SE LLAMA ORIGINAL ORDER!!!!!!

Como esta columna es unicamente para referenciar registros, vamos a ordenarla de 1 a 6312. Pero necesitamos saber como voy a tratar sus 3 nulos.

In [ ]:
sharks[sharks['original order'].isna()]

Claramente estos 3 registros no aportan valor. Procedemos a eliminarlos

In [ ]:
bad_index = sharks[sharks['original order'].isna()].index
sharks.drop(index=bad_index,inplace=True)

Comlpetamos esta columna en el rango de 1 a n. Esto nos puede dar información más adelante para completar fechas o registros

In [ ]:
sharks['original order'].is_monotonic_increasing

In [ ]:
sharks.loc[600:630,['original order']]

In [ ]:
sharks.head(20)